pkg/virtual/framework/rootapiserver/root_apiserver.go

```diff
@@ -210,12 +212,36 @@ func (c completedConfig) NewRequestInfo(req *http.Request) (*genericapirequest.R
 	return defaultResolver.NewRequestInfo(req)
 }
 
-func NewRootAPIConfig(secureServing *genericapiserveroptions.SecureServingOptionsWithLoopback, authenticationOptions *genericapiserveroptions.DelegatingAuthenticationOptions, informerStarts InformerStarts, virtualWorkspaces ...framework.VirtualWorkspace) (*RootAPIConfig, error) {
+func NewRootAPIConfig(secureServing *genericapiserveroptions.SecureServingOptions, authenticationOptions *genericapiserveroptions.DelegatingAuthenticationOptions, informerStarts InformerStarts, virtualWorkspaces ...framework.VirtualWorkspace) (*RootAPIConfig, error) {
 	genericConfig := genericapiserver.NewRecommendedConfig(legacyscheme.Codecs)
 
 	// TODO: genericConfig.ExternalAddress = ... allow a command line flag or it to be overridden by a top-level multiroot apiServer
 
-	if err := secureServing.ApplyTo(&genericConfig.Config.SecureServing, &genericConfig.Config.LoopbackClientConfig); err != nil {
+	if err := secureServing.ApplyTo(&genericConfig.Config.SecureServing); err != nil {
+		return nil, err
+	}
+
+	var err error
+
+	var loopbackConfigCert []byte
+	if genericConfig.Config.SecureServing.Cert != nil {
+		loopbackConfigCert, _ = genericConfig.Config.SecureServing.Cert.CurrentCertKeyContent()
+	}
+	for _, sniCert := range genericConfig.Config.SecureServing.SNICerts {
+		if loopbackConfigCert == nil {
+			loopbackConfigCert, _ = sniCert.CurrentCertKeyContent()
+		}
+		if sets.NewString(sniCert.SNINames()...).Has(genericapiserver.LoopbackClientServerNameOverride) {
+			loopbackConfigCert, _ = sniCert.CurrentCertKeyContent()
+		}
+	}
+
+	if loopbackConfigCert == nil {
+		return nil, errors.New("No certs setup")
+	}
+
+	genericConfig.Config.LoopbackClientConfig, err = genericConfig.Config.SecureServing.NewLoopbackClientConfig(uuid.New().String(), loopbackConfigCert)
+	if err != nil {
```

Comment:
are you sure this loopback client does not exist already? (I don't remember either, but could well be).

Reply:
I don't think it exists since we didn't use `SecureServingOptionsWithLoopback`.
In fact the loopback we create here is just because one is required, but the RootAPIServer wouldn't really need one since it doesn't manages REST objects on it own, and the virtual workspaces it hosts will generally have their own config that will point to an external delegate APIServer (typically KCP, or a proxy, or another virtual workspace).

Reply:
not sure I understand this change then. Why do we need the new code?

Reply:
Because we should no run `SecureServingOptionsWithLoopback.ApplyTo`, because it creates and additional Lookpack config with its own SNIcerts, etc ... which we don't want.